### PR TITLE
feat(vercel): add OTEL based LangSmith trace exporter

### PR DESCRIPTION
**Pull request by dqbd**
Add OTEL-based LangSmith trace exporter, conforming to Vercel AI SDK Telemetry documentation. 

TODO:
- [x] Add unit tests with mock results and mock language model to assert the outputs as well
- [x] Verify how wrapAISDKModel would work with LangSmithAISDKExporter 
- [x] Add `first_token_time`

### CODE REVIEW

File: js/src/tests/vercel_exporter.int.test.ts [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+import { Client } from "../index.js";
+import { waitUntilRunFound } from "./utils.js";
+
+const getTelemetrySettings = (langsmithRunId?: string) => {
+  const metadata: Record<string, string> = {
+    userId: "123",
+    language: "english",
+  };
+
+  if (langsmithRunId) metadata.langsmithRunId = langsmithRunId;
```
Comment (dqbd) [inline js/src/tests/vercel_exporter.int.test.ts:27]: This method allows specifying the `runId` beforehand, open to different naming.

File: js/src/wrappers/vercel/exporter.ts [hunk 2]
```diff
@@ -0,0 +1,647 @@
...
+  } catch {
+    return str;
+  }
+};
+
+function stripNonAlphanumeric(input: string) {
+  return input.replace(/[-:.]/g, "");
+}
+
+function convertToDottedOrderFormat(
```
Comment (dqbd) [inline js/src/wrappers/vercel/exporter.ts:181]: AFAICT, the order in which the spans arrive to the trace exporter matches the "execution order",

File: js/src/wrappers/vercel/exporter.ts [hunk 3]
```diff
@@ -0,0 +1,647 @@
...
+
+    if (incompleteNodes.length > 0) {
+      console.warn(
+        "Some incomplete nodes were found before shutdown and not sent to LangSmith."
+      );
+    }
+
+    await this.client?.awaitPendingTraceBatches();
+  }
+  async forceFlush?(): Promise<void> {
```
[the remark this answers is not in the document]
Comment (dqbd) [reply js/src/wrappers/vercel/exporter.ts:644]: Part of OTEL

File: js/src/tests/vercel_exporter.int.test.ts [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+    isEnabled: true,
+    functionId: "functionId",
+    metadata,
+  };
+};
+
+const client = new Client();
+// Not using @opentelemetry/sdk-node because we need to force flush
+// the spans to ensure they are sent to LangSmith between tests
+const provider = new NodeTracerProvider();
```
[the remark this answers is not in the document]
Comment (dqbd) [reply js/src/tests/vercel_exporter.int.test.ts:38]: To only exclude certain runs you can exclude the `experimental_telemetry` prop from `stream/generate` method.

File: js/src/wrappers/vercel/exporter.types.ts [hunk 5]
```diff
@@ -0,0 +1,230 @@
+import type { ReadableSpan } from "@opentelemetry/sdk-trace-base";
```
[the remark this answers is not in the document]
Comment (dqbd) [reply js/src/wrappers/vercel/exporter.types.ts:1]: Made the external type definition to accept `undefined`, which should omit the type during build.

[the remark this answers is not in the document]
Comment (dqbd) [reply js/src/wrappers/vercel/exporter.types.ts:1]: Also added `stripInternal` to hide internal methods from type definitions, which should result in empty index/typedef.

File: js/src/wrappers/vercel/exporter.ts [hunk 6]
```diff
@@ -0,0 +1,697 @@
...
+import { KVMap, RunCreate } from "../../schemas.js";
+import { v5 as uuid5 } from "uuid";
+
+// eslint-disable-next-line @typescript-eslint/ban-types
+type AnyString = string & {};
+
+type LangChainMessageFields = {
+  content:
+    | string
+    | Array<
```
Comment (jacoblee93) [inline js/src/wrappers/vercel/exporter.ts:13]: Are we typing this too strongly?

File: js/src/wrappers/vercel/exporter.ts [hunk 7]
```diff
@@ -0,0 +1,697 @@
...
+        return part;
+      });
+
+      const toolCalls = message.content.filter(
+        (part): part is ToolCallPart => part.type === "tool-call"
+      );
+
+      if (toolCalls.length > 0) {
+        data.additional_kwargs ??= {};
+        data.additional_kwargs.tool_calls = toolCalls.map((part) => {
```
Comment (jacoblee93) [inline js/src/wrappers/vercel/exporter.ts:85]: Shouldn't we just populate `tool_calls` on messages?

File: js/src/wrappers/vercel/exporter.ts [hunk 8]
```diff
@@ -0,0 +1,697 @@
...
+                };
+          }
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        | (Record<string, any> & { type?: "text" | "image_url" | string })
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        | (Record<string, any> & { type?: never })
+      >;
+  name?: string;
+  id?: string;
+  additional_kwargs?: {
```
Comment (jacoblee93) [inline js/src/wrappers/vercel/exporter.ts:31]: `tool_calls` is top level now?

File: js/src/wrappers/vercel/exporter.ts (hunk 6 above)
Comment (dqbd) [reply js/src/wrappers/vercel/exporter.ts:13]: This actually relaxes the string type to permit auto-complete while allowing other strings

File: js/src/wrappers/vercel/exporter.ts [hunk 9]
```diff
@@ -0,0 +1,697 @@
...
+      relativeExecutionOrder: Record<string, number>;
+      interop?: InteropType;
+    }
+  > = {};
+
+  constructor(args?: { client?: Client }) {
+    this.client = args?.client ?? new Client();
+  }
+
+  /** @internal */
```
Comment (jacoblee93) [inline js/src/wrappers/vercel/exporter.ts:244]: nit: I actually don't think this does anything - would prefix with `_` instead

File: js/src/wrappers/vercel/exporter.ts [hunk 10]
```diff
@@ -0,0 +1,697 @@
...
+            result.llm_output ??= {};
+            result.llm_output.token_usage ??= {};
+            result.llm_output.token_usage["prompt_tokens"] =
+              span.attributes["ai.usage.promptTokens"];
+          }
+
+          return result;
+        })();
+
+        // TODO: add first_token_time
```
Comment (jacoblee93) [inline js/src/wrappers/vercel/exporter.ts:410]: I do think we will want this, folks have asked. Fine to follow up later

File: js/src/wrappers/vercel/exporter.ts [hunk 11]
```diff
@@ -0,0 +1,697 @@
...
+      const parentRunId = parentId
+        ? uuid5(parentId, RUN_ID_NAMESPACE)
+        : undefined;
+
+      const traceMap = this.traceByMap[traceId];
+
+      const run = this.getRunCreate(span);
+      if (!run) continue;
+
+      traceMap.relativeExecutionOrder[parentRunId ?? "$"] ??= -1;
```
Comment (jacoblee93) [inline js/src/wrappers/vercel/exporter.ts:557]: nit: make `$` a constant? Or is this an OTEL thing

**Review by jacoblee93 — APPROVED** [review]

**Pull request merged**